File: Assets/Scripts/Controller/ScoreController.java
package Controller;

import Model.BoardModel;
import Model.GameStateModel;
import Model.PlayerModel;
import Model.TurnCounterModel;

public class ScoreController {
    GameStateModel currentGameStateModel;
    BoardModel gameBoardModel;
    PlayerModel currentPlayerModel;
    TurnCounterModel Counter;
    char[][] Board;

    int XMoves;
    int OMoves;
    int XWins;
    int OWins;

    public GameStateModel CheckCurrentGameState(GameStateModel currentGameStateModel, PlayerModel PlayerModel, BoardModel gameBoardModel, TurnCounterModel Counter){
        this.currentGameStateModel = currentGameStateModel;
        this.currentPlayerModel = PlayerModel;
        this.gameBoardModel = gameBoardModel;
        this.Counter = Counter;
        Board = gameBoardModel.GetFullBoard();
        int PlayersTotalMoves;

        CountWins();

        XMoves = PlayerModel.GetMoves(PlayerModel.X);
        OMoves = PlayerModel.GetMoves(PlayerModel.O);
        XWins  = PlayerModel.GetWins(PlayerModel.X);
        OWins  = PlayerModel.GetWins(PlayerModel.O);
        PlayersTotalMoves = XMoves + OMoves;

        if (XWins == 0 && OWins > 0){           currentGameStateModel = GameStateModel.O_Win;
        } else if(XWins >  0 && OWins == 0){    currentGameStateModel = GameStateModel.X_Win;
        } else if(XWins == 0 && OWins == 0){
            if(PlayersTotalMoves < 9){          currentGameStateModel = GameStateModel.Playing;
            } else if (PlayersTotalMoves == 9){ currentGameStateModel = GameStateModel.Draw;
            }
        }
        return currentGameStateModel;
    }

    public void CountWins(){
        for (int i = 0; i < 3; i++) { // Verticals
            if (Board[i][0] + Board[i][1] + Board[i][2] == 264) currentPlayerModel.AddWins(PlayerModel.X);
            if (Board[i][0] + Board[i][1] + Board[i][2] == 237) currentPlayerModel.AddWins(PlayerModel.O);
        }

        for (int i = 0; i < 3; i++) { // Horizontals
            if (Board[0][i] + Board[1][i] + Board[2][i] == 264) currentPlayerModel.AddWins(PlayerModel.X);
            if (Board[0][i] + Board[1][i] + Board[2][i] == 237) currentPlayerModel.AddWins(PlayerModel.O);
        }

        // Diagonals
        if (Board[0][0] + Board[1][1] + Board[2][2] == 264) currentPlayerModel.AddWins(PlayerModel.X);
        if (Board[0][0] + Board[1][1] + Board[2][2] == 237) currentPlayerModel.AddWins(PlayerModel.O);
        if (Board[0][2] + Board[1][1] + Board[2][0] == 264) currentPlayerModel.AddWins(PlayerModel.X);
        if (Board[0][2] + Board[1][1] + Board[2][0] == 237) currentPlayerModel.AddWins(PlayerModel.O);
    }
}

File: Assets/Scripts/Controller/ScoreController.cs
using Model;
using UnityEngine;

namespace Controller
{
    public class ScoreController : MonoBehaviour
    {
        public GameStateEnum currentGameState = GameStateEnum.Playing;
        private const int playerXID = 0;
        private const int playerOID = 1;
        private char[,] xMoves;
        private char[,] oMoves;
        private int xWins;
        private int oWins;
        private int TotalMovesPlayed;
        
        public GameStateEnum CheckCurrentGameState(GameObject playersModel, PlayerEnum currentPlayer)
        {
            CountMoves(playersModel);
            CountWins(playersModel);
            
            xWins = playersModel.GetComponent<PlayersModel>().GetWins(playerXID);
            oWins = playersModel.GetComponent<PlayersModel>().GetWins(playerOID);
            TotalMovesPlayed = playersModel.GetComponent<PlayersModel>().GetTotalMoves();
            
            if (xWins == 0 && oWins > 0){           currentGameState = GameStateEnum.OWin;
            } else if(xWins >  0 && oWins == 0){    currentGameState = GameStateEnum.XWin;
            } else if(xWins == 0 && oWins == 0){
                if(TotalMovesPlayed < 9){          currentGameState = GameStateEnum.Playing;
                } else if (TotalMovesPlayed == 9){ currentGameState = GameStateEnum.Draw;
                }
            }
            return currentGameState;
        }
        
        void CountMoves (GameObject playersModel)
        {
            xMoves = playersModel.GetComponent<PlayersModel>().GetMoves(playerXID);
            oMoves = playersModel.GetComponent<PlayersModel>().GetMoves(playerOID);
        }

        void CountWins(GameObject playersModel)
        {
            for (int i = 0; i < 3; i++) { // Verticals
                if (xMoves[i,0] + xMoves[i,1] + xMoves[i,2] == 264)
                    playersModel.GetComponent<PlayersModel>().AddWins(playerXID);
                if (oMoves[i,0] + oMoves[i,1] + oMoves[i,2] == 237) 
                    playersModel.GetComponent<PlayersModel>().AddWins(playerOID);
            }

            for (int i = 0; i < 3; i++) { // Horizontals
                if (xMoves[0,i] + xMoves[1,i] + xMoves[2,i] == 264)
                    playersModel.GetComponent<PlayersModel>().AddWins(playerXID);
                if (oMoves[0,i] + oMoves[1,i] + oMoves[2,i] == 237) 
                    playersModel.GetComponent<PlayersModel>().AddWins(playerOID);
            }

            // Diagonals
            if (xMoves[0,0] + xMoves[1,1] + xMoves[2,2] == 264) playersModel.GetComponent<PlayersModel>().AddWins(playerXID);
            if (oMoves[0,0] + oMoves[1,1] + oMoves[2,2] == 237) playersModel.GetComponent<PlayersModel>().AddWins(playerOID);
            if (xMoves[0,2] + xMoves[1,1] + xMoves[2,0] == 264) playersModel.GetComponent<PlayersModel>().AddWins(playerXID);
            if (oMoves[0,2] + oMoves[1,1] + oMoves[2,0] == 237) playersModel.GetComponent<PlayersModel>().AddWins(playerOID);
        }
    }
}
